llvm/lib/Target/SPIRV/SPIRVGlobalTypesAndRegNumPass.cpp

```diff
@@ -12,15 +12,16 @@
 // Before this pass, all SPIR-V instructions were local scope, and registers
 // were numbered function-locally. However, SPIR-V requires Type instructions,
 // global variables, capabilities, annotations etc. to be in global scope and
-// occur at the start of the file. This pass hoists these as necessary.
+// occur at the start of the file. This pass copies these as necessary to dummy
+// function which represents the global scope. The original local instructions
+// are left to keep MIR consistent. They will not be emitted to asm/object file.
 //
-// This pass also re-numbers the registers globally, and patches up any
-// references to previously local registers that were hoisted, or function IDs
-// which require globally scoped registers.
+// This pass also re-numbers the registers globally, and saves global aliases of
+// local registers to the global registry. AsmPrinter uses these aliases to
+// output instructions that refers global registers.
 //
-// This pass breaks all notion of register def/use, and generated MachineInstrs
-// that are technically invalid as a result. As such, it must be the last pass,
-// and requires instruction verification to be disabled afterwards.
+// In OpFunctionCalls callee identifiers are substituted with function IDs
```

Comment:
Why aliases won't work for `OpFunctionCall`s as well?

Reply:
There were 2 reasons: 
- we have no such alias in RegisterAliasTable[MF] because OpFunction is defined outside of current MF; that's why we build FuncNameToID table in assignFunctionCallIDs.
- it looks like we need to recreate each OpFunctionCall to change callee GlobalIdentifier string to FuncID register; at that place we already have global FuncID register, so it seems natural to directly use it instead of making a new local FuncID register and adding corresponding alias to RegisterAliasTable[MF] for later use in AsmPrinter.

But using "global" register may break MIR consistency, so probably we should generate a code with local registers like this:
```
%newLocalReg = @foo // setSkipEmission
%reg = OpFunctionCall %type %newLocalReg ... // setRegisterAlias(MF, newLocalReg, globalFuncID)
```
I'll add TODO for this change.

Reply:
As `%newLocalReg = @foo` we could use OpFunction (closer to IR/SPIR-V logic) or even OpConstantNull (easier to generate). Both will be invalid SPIR-V code but MIR should be consistent.

Reply:
> * we have no such alias in RegisterAliasTable[MF] because OpFunction is defined outside of current MF; that's why we build FuncNameToID table in assignFunctionCallIDs.

What do you mean exactly? Any MF contains its `OpFunction` just as it contains types/consts/etc

> * it looks like we need to recreate each OpFunctionCall to change callee GlobalIdentifier string to FuncID register; at that place we already have global FuncID register, so it seems natural to directly use it instead of making a new local FuncID register and adding corresponding alias to RegisterAliasTable[MF] for later use in AsmPrinter.

I see. In fact having a `global` operand in `OpFunctionCall` conflicts with its description in `InstrInfo.td`, not sure why it's not a verification failure.

What I'm actually talking about is what if we unify the approach to handle globals for functions too. What I mean - let's assume funcs `f1` and f2 both call `f3`. Then both `f1` and `f2` (and `f3` as its actual definition) should have its own copy of `OpFunction` which is going to be tracked in the existing DT and then deduplicated as other globals/constants. Makes sense?

Reply:
> What do you mean exactly? Any MF contains its OpFunction just as it contains types/consts/etc

I mean that currently MF with OpFunctionCall has no local OpFunction instruction for the called function.

>What I'm actually talking about is what if we unify the approach to handle globals for functions too...

It's a good suggestion for TODO in CallLowering/GlobalTypesAndRegNumPass (and maybe it requires changes in passes between these two). For now I have implemented a simple fix in assignFunctionCallIDs (SPIRVGlobalTypesAndRegNumPass.cpp):  the local OpFunction (actually it is dummy) is inserted just before OpFunctionCall and correspondent register alias is created there. With this fix the check `MI->getOpcode() == SPIRV::OpFunctionCall && i == 2` is not required in SPIRVMCInstLower.cpp. Once testing for pr103 is positively passed, I'll send this patch for review.

> PS For parameters there might be variants:

I think we don't need to insert OpFunctionParameter inside a caller functions because their results aren't used in the caller and OpFunctionParameters make sense only in definition/declaration of a function itself.